uwblueprint/focus-on-nature
# feat: bulk edit  and delete campers

## Notion ticket link

[Removing and moving campers in bulk](https://www.notion.so/uwblueprintexecs/Removing-and-moving-campers-in-bulk-e72e226660ca46688ce7ec3c0ca1a61e)



## Implementation description
* `updateCampersById` → takes in an array of `camperId`s and an `UpdateCamperDTO` that specifies the fields that need to be updated for all the campers specified (via the `camperId`s). Update happens as long as all campers have the same `chargeId`. If `campSession` is passed as a field to be updated, then the `campers` array in both the new and old `campSession`s are updated.
* Note: if want to update distinct fields for different campers, will have to call this update function for each camper/fields to be updated
* `deleteCampersById` → can take in an arbitrary number of `camperId`s, and as long as they are all existing campers with the same `chargeId` then they can be deleted
* `cancelRegistration` → can take in an arbitrary number of `camperId`s and a `chargeId` associated with all the campers, then does a check for whether the camp session that these campers are registered for starts more than 30 days after the cancellation OR if the waitlist for the camp is not empty and is less than 30 days from cancellation to start date. Then calls the `deleteCampersById` function if these checks pass
* Assumption for all these functions: `chargeId` is the parameter that establishes that these campers are all from the same family (can be edited/delete together) - this may need to change later



## Steps to test `updateCampersById`
1. Use this testing endpoint on Postman: http://localhost:5000/campers/
2. Pass in `req.body`: an array of `camperId`s under `camperIds` and the camper fields that need to be updated (e.g. {"camperIds": [...], "campSession": ...}). Format like this
```
{
    "camperIds": [
        "625786928d4e72b0e41193a6",
        "62576da892c3a430022c653d"
    ],
    "firstName": "JANET",
    // add any other fields to be updated

}
```
3. Test cases: updating one or many campers (making sure the same fields are updated), omitting certain fields in the update DTO, testing that the update still works and testing that the `camper` array in the`campSession`s are also updated if that field is passed
4. Error cases (these cases should throw errors): moving campers between `campSession`s that are not under the same camp, wrong primitive values passed, `camperId`s for campers that don’t exist, `campSession` that doesn’t exist, different `chargeId`s on each camper, invalid `campSession`

## Steps to test `deleteCampersById`
1. Use this testing endpoint on Postman: http://localhost:5000/campers/
2. Pass in `req.body`: an array of `camperId`s (e.g. "camperIds" : ["`6248a739dd79f2313c922c85`", "`6248a73add79f2313c922c8a`"])
3. Test cases: deleting 0, 1, 2+ campers, checking that the campers are removed from the `campSession`'s array of `campers`
4. Error cases (these cases should throw errors): campers who don’t have the same `chargeId`, campers that don’t exist in the DB

## Steps to test `cancelRegistration`
1. Use this testing endpoint on Postman: http://localhost:5000/campers/cancel-registration
2. Pass in `req.body`: an array of `camperId`s (e.g. "camperIds" : ["`6248a739dd79f2313c922c85`", "`6248a73add79f2313c922c8a`"]) and a `chargeId`
3. Test cases: cancelling registration for 0, 1, 2+ campers, cancelling registration if the `campSession` start date > 30 days or waitlist is not empty, check that refund is made on stripe and that campers are removed from the `campSession`'s array of `campers`
4. Error cases (these cases should throw errors): campers that are not all registered for the same `campSession`, campers that don’t exist in the DB, camp session start date < 30 days and waitlist is empty



## What should reviewers focus on?
* Does the code make sense for the use cases it is made for?


## Checklist
